src/Donations/Migrations/RecalculateExchangeRate.php

```diff
@@ -0,0 +1,162 @@
+    }
+
+    /**
+     * @inheritDoc
+     * @throws DatabaseMigrationException
+     */
+    public function runBatch($firstId, $lastId): void
+    {
+        try {
+            // Select donations where base currency is not empty,
+            // base currency is different from payment currency,
+            // and either the exchange rate does not exist or is equal to 1
+
+            $query = $this->query()
+                ->select('posts.ID')
+                ->attachMeta(
+                    'give_donationmeta',
+                    'ID',
+                    'donation_id',
+                    [DonationMetaKeys::AMOUNT, 'amount'],
+                    [DonationMetaKeys::CURRENCY, 'currency'],
+                    [DonationMetaKeys::BASE_AMOUNT, 'baseAmount'],
+                    [DonationMetaKeys::EXCHANGE_RATE, 'exchangeRate'],
+                    ['_give_cs_base_currency', 'baseCurrency']
+                )
+                ->whereIsNotNull('give_donationmeta_attach_meta_baseAmount.meta_value')
+                ->whereRaw('AND give_donationmeta_attach_meta_baseCurrency.meta_value != give_donationmeta_attach_meta_currency.meta_value')
+                ->where(function ($query) {
+                    $query->where('give_donationmeta_attach_meta_exchangeRate.meta_value', '1')
+                        ->orWhereIsNull('give_donationmeta_attach_meta_exchangeRate.meta_value');
+                });
+
+            // Migration Runner will pass null for lastId in the last step
+            if (is_null($lastId)) {
+                $query->where('posts.ID', $firstId, '>');
+            } else {
+                $query->whereBetween('posts.ID', $firstId, $lastId);
+            }
+
+            $donations = $query->getAll();
+
+            foreach ($donations as $donation) {
+                $amount = Money::fromDecimal($donation->amount, $donation->currency);
+                $baseAmount = Money::fromDecimal($donation->baseAmount, $donation->baseCurrency);
+
+                if ($baseAmount->formatToDecimal() === '0') {
+                    continue;
+                }
+
+                /** @var Money $exchangeRate */
+                $exchangeRate = $amount->divide($baseAmount->formatToDecimal());
+
+                give()->payment_meta->update_meta($donation->ID, DonationMetaKeys::EXCHANGE_RATE, $exchangeRate->formatToDecimal());
+            }
+        } catch (DatabaseQueryException $exception) {
+            throw new DatabaseMigrationException(
+                'An error occurred while updating donation exchange rates',
+                0,
+                $exception
+            );
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getItemsCount(): int
+    {
+        return $this->query()->count();
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getBatchItemsAfter($lastId): ?array
+    {
+        $item = $this->query()
+            ->select('MIN(ID) AS first_id, MAX(ID) AS last_id')
```

Comment:
This won't work. To get the min ID and max ID, you have to use a sub-select query to get the range of IDs, then you can select the min and max ID from that range.

Example:
```sql
SELECT MIN(ID) AS first_id, MAX(ID) AS last_id FROM (SELECT ID FROM table WHERE ID > someID ORDER BY ID ASC LIMIT batchSize)
```

Reply:
I don't think there's a problem there but lemme know what you think. Here is some tests I did before rewriting that query.

![CleanShot 2025-04-25 at 10 40 35](https://github.com/user-attachments/assets/5e810a2c-c80b-4bae-bf94-af4a22991f67)

![CleanShot 2025-04-25 at 10 41 17](https://github.com/user-attachments/assets/6265791a-84b2-4233-85b7-74d9a0e9f41d)

Reply:
Here is the test using your SQL query:
![CleanShot 2025-04-25 at 10 44 18](https://github.com/user-attachments/assets/9cb9e37d-1536-4433-956f-1d9449d4ef61)

Same results.

Reply:
What is the last ID in database? Without limit

Reply:
I see what's the issue here. Limit is applied to the number of rows being returned, which in this case is just 1, so it does not applies to trim the range of IDs.

Reply:
Updated on https://github.com/impress-org/givewp/pull/7893/commits/ccfb90da152727c9f4fb3ce25afaee2e2faa6962